text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,208 @@
+
+# Proposal
+
+First, a new table called "recording_byte_range" with the following structure is created:
+
+| replay_id | segment_id | path     | start | stop  |
+| --------- | ---------- | -------- | ----- | ----- |
+| A         | 0          | file.bin | 0     | 6241  |
+| B         | 0          | file.bin | 6242  | 8213  |
+| A         | 1          | file.bin | 8214  | 12457 |
+
+Replay and segment ID should be self-explanatory. Path is the location of the blob in our bucket. Start and stop are integers which represent the index positions in an inclusive range. This range is a contiguous sequence of related bytes. In other words, the segment's compressed data is contained within the range.
+
+Notice each row in the example above points to the same file but with different start and stop locations. This is implies that multiple segments and replays can be present in the same file. A single file can be shared by hundreds of different segments.
+
+This table will need to support, at a minimum, one write per segment.
+
+Second, the Session Replay recording consumer will not commit blob data to Google Cloud Storage for each segment. Instead it will buffer many segments and flush them together as a single blob to GCS. Next it will make a bulk insertion into the database for tracking.
+
+```mermaid
+flowchart
+    A[Wait For New Message] --> B[Process];
+    B --> C[Push to Buffer];
+    C --> D{Buffer Full?};
+    D -- No --> A;
+    D -- Yes --> E[Write Single File to GCS];
+    E --> F[Bulk Insert Byte Ranges];
+    F --> G[Clear Buffer];
+    G --> A;
+```
+
+Third, when a client requests recording data we will look it up in the "recording_byte_range" table. From it's response, we will issue as many fetch requests as there are rows in the response. These requests may target a single file or many files. The files will be fetched with a special header that instructs the service provider to only respond with a subset of the bytes. Specifically, the bytes that related to our replay.
+
+The response bytes will be decompressed, merged into a single payload, and returned to the user as they are now.
+
+# Drawbacks
+
+1. Deleting data becomes tricky. See "Unresolved Questions".
+
+# Unresolved Questions
+
+1. Can we keep deleted data in GCS but make it inaccessible?
+
+   - User and project deletes:
+     - We would remove all capability to access it making it functionally deleted.
+   - GDPR deletes:
+     - Would this require downloading the file, over-writing the subsequence of bytes, and re-uploading a new file?
+       - Single replays, small projects, or if the mechanism is infrequently used could make this a valid deletion mechanism.
+       - The data could be encrypted, with some encryption key stored on the metadata row, making the byte sequence unreadable upon row delete.
+
+2. What datastore should we use to store the byte range information?
+
+   - Cassandra, Postgres, AlloyDB?
+   - Postgres likely won't be able to keep up long-term.
```

Comment:
Do you have any numbers to back up that claim? postgres can do a lot, espcially if yall get a dedicated postgres instance for this

Reply:
oh yes we should use AlloyDB for this (it is just postgres, but managed meaning the likelihood of getting paged for it is much less).

Reply:
Perfect 👍